Documentation/DesignDocs/corewcfqueue-design.md

```diff
@@ -0,0 +1,308 @@
+    private readonly IServiceScopeFactory _servicesScopeFactory;
+
+    public QueueHandShakeMiddleWare(
+        IServiceProvider services,
+        IServiceScopeFactory servicesScopeFactory, IQueueMiddlewareBuilder queueBuilder)
+    {
+        _services = services;
+        _servicesScopeFactory = servicesScopeFactory;
+        _queueMiddlewareBuilder = InitQueueMiddleWare(queueBuilder);
+    }
+
+    private IQueueMiddlewareBuilder InitQueueMiddleWare(IQueueMiddlewareBuilder queueBuilder)
+    {
+        queueBuilder.UseMiddleware<QueueFetchMessage>();
+        queueBuilder.UseMiddleware<QueueProcessMessage>();
+        return queueBuilder;
+    }
+
+    public QueueMessageDispatcherDelegate Build()
+    {
+        return _queueMiddlewareBuilder.Build();
+    }
+       
+}
+```
+#### IQueueTransport, QueueBaseTransportBindingElement , QueueTransportContext and QueueTransportPump
+- `IQueueTransport` defines a single method, `ReceiveQueueMessageContextAsync`. The `QueuePollingService` will call this method to fetch the next incoming queue message. 
+```csharp
+   public interface IQueueTransport
+    {
+        Task<QueueMessageContext> ReceiveQueueMessageContextAsync(CancellationToken cancellationToken);
+    }
+```
+- `QueueBaseTransportBindingElement` is an abstract class with `BuildQueueTransportPump` as an abstract method
+- Each Queue transport binding element must extend `QueueBaseTransportBindingElement` and implement `BuildQueueTransportPump`.
+```csharp
+    public abstract class QueueBaseTransportBindingElement : TransportBindingElement
+    {
+        protected QueueBaseTransportBindingElement()
+        {
+        }
+
+        protected QueueBaseTransportBindingElement(QueueBaseTransportBindingElement elementToBeCloned)
+            : base(elementToBeCloned)
+        {
+        }
+
+        public virtual int MaxPendingReceives { get { return 1; } }
+        public abstract QueueTransportPump BuildQueueTransportPump(BindingContext context);
+
+        public override bool CanBuildServiceDispatcher<TChannel>(BindingContext context)
+        {
+            return (typeof(TChannel) == typeof(IInputChannel));
+        }
+
+        public override IServiceDispatcher BuildServiceDispatcher<TChannel>(BindingContext context, IServiceDispatcher innerDispatcher)
+        {
+            return innerDispatcher;
+        }
+
+    }
+ ```
+- The concrete implementation of `IQueueTransport` will be used by CoreWCF to retrieve messages from the queue storage service. 
+- When `ReceiveQueueMessageContextAsync` method is called, it returns an instance of `QueueMessageContext` object.
+- `ServiceDispatcher`, `QueueBaseTransportBindingElement`, `QueueMessageDispatcherDelegate`, and `QueueTransportPump` are all stored in `QueueTransportContext`. 
+-`QueueTransportPump` is an abstract class with `StartPumpAsync` and `StopPumpAsync` task. 
+- Any queue transport which is using push mechanism (e:g:- RabbitMQ) needs to implement this class.
```

Comment:
Inconsistent spacing between - and text on this line. Also, StartPumpAsync and StopPumpAsync are methods, not tasks. Maybe refer to them as "Task returning methods". Technically calling them async methods isn't accurate as they could be implemented synchronously, they just need to return a Task.